server/mongo_indexes.go

```diff
@@ -0,0 +1,199 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"strings"
+
+	mgo "gopkg.in/mgo.v2"
+)
+
+// IndexMap is a map of index arrays with the collection name as the key. Each index array
+// contains one or more *mgo.Index indexes.
+type IndexMap map[string][]*mgo.Index
+
+// ConfigureIndexes ensures that all indexes listed in the provided indexes.conf file
+// are part of the Mongodb fhir database. If an index does not exist yet ConfigureIndexes
+// creates a new index in the background using mgo.collection.EnsureIndex(). Depending
+// on the size of the collection it may take some time before the index is created.
+// This will block the current thread until the indexing completes, but will not block
+// other connections to the mongo database.
+func ConfigureIndexes(session *mgo.Session, config Config) {
+	var err error
+
+	db := session.DB(config.DatabaseName)
+
+	// Read the config file
+	idxConfig, err := ioutil.ReadFile(config.IndexConfigPath)
+	if err != nil {
+		log.Println("[WARNING] Could not find indexes configuration file")
+	}
+
+	// parse the config file
+	indexMap, err := parseIndexes(string(idxConfig))
+
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	// ensure all indexes in the config file
+	for k := range indexMap {
+		collection := db.C(k)
+
+		for _, index := range indexMap[k] {
+			log.Printf("Ensuring index: %s.%s: %s\n", config.DatabaseName, k, sprintIndexKeys(index))
+			err = collection.EnsureIndex(*index)
+
+			if err != nil {
+				log.Printf("[WARNING] Could not ensure index: %s.%s: %s\n", config.DatabaseName, k, sprintIndexKeys(index))
+			}
+		}
+	}
+	session.Close()
```

Comment:
The session should not be closed here.  Generally, a good rule of thumb is that whatever function _creates_ / _opens_ a resource, that function is also responsible for _closing_ it.  So, I'd do one of the following:

1) In `server_setup.go`, pass the main session to `ConfigureIndexes`, and then `ConfigureIndexes` should create a copy of the session and close it when it's done (using a `defer` to ensure that any early exits, like `log.Fatal()`, still close the session).

or

2) Have `server_setup.go` copy the session, pass in the copied session to `ConfigureIndexes` and then close the copied session after `ConfigureIndexes` returns.